### PR TITLE
Add a note to the ArbitraryExpressionInPattern error

### LINKED ISSUE #99380: Confusing "arbitrary expressions aren't allowed in patterns" error
Take the following code:

```rust
macro_rules! foo {
    ($p:expr) => {
        if let $p = Some(42) {
            return;
        }
    };
}

fn main() {
    foo!(Some(3));
}
```

Currently it fails with:

```
error: arbitrary expressions aren't allowed in patterns
  --> src/main.rs:10:10
   |
10 |     foo!(Some(3));
   |          ^^^^^^^
```

If you expanded the macro manually, `Some(3)` would be parsed as a pattern, but the `$p:expr` above changes it to be parsed as an expression. So what you *need* to do here is to change the `$p:expr` above to `$p:pat`. But this is very non-obvious.

Ideally, the compiler would try to check if the input is a valid pattern, and if so, add a note like "`Some(3)` is a valid pattern but the metavariable $p is specified to be an expression", and maybe also point to the metavariable's matcher.

@rustbot label A-macros

### PR DESCRIPTION
**Pull request by est31**
The current "arbitrary expressions aren't allowed in patterns" error is confusing, as it fires for code where it *looks* like a pattern but the compiler still treats it as an expression. That this is due to the `:expr` fragment specifier forcing the expression-ness property on the code.

In the test suite, the "arbitrary expressions aren't allowed in patterns" error can only be found in combination with macro_rules macros that force expression-ness of their content, namely via `:expr` metavariables. I also can't come up with cases where there would be an expression instead of a pattern, so I think it's always coming from an `:expr`.

In order to make the error less confusing, this adds a note explaining the weird `:expr` fragment behaviour.

Fixes #99380

### CODE REVIEW

File: compiler/rustc_ast_lowering/messages.ftl [hunk 1]
```diff
@@ -5,6 +5,7 @@ ast_lowering_abi_specified_multiple_times =
 
 ast_lowering_arbitrary_expression_in_pattern =
     arbitrary expressions aren't allowed in patterns
+    .pattern_from_macro_note = the :expr fragment specifier forces the metavariable's content to be an expression
```
Comment (fmease) [inline compiler/rustc_ast_lowering/messages.ftl:8]:
```suggestion
    .pattern_from_macro_note = the `expr` fragment specifier forces the metavariable's content to be an expression
```

Comment (est31) [reply compiler/rustc_ast_lowering/messages.ftl]: It seems the `invalid fragment specifier` error also uses backticks:

```
error: invalid fragment specifier `r#if`
  --> $DIR/invalid-fragment-specifier.rs:7:6
   |
LL |     ($wrong:r#if) => {};
   |      ^^^^^^^^^^^
   |
   = help: valid fragment specifiers are `ident`, `block`, `stmt`, `expr`, `pat`, `ty`, `lifetime`, `literal`, `path`, `meta`, `tt`, `item` and `vis`
```

so switching to those.

**Comment by est31** [comment]
> I also can't come up with cases where there would be an expression instead of a pattern

Note that should in the future anyone discover cases where there was no involvement of `:expr` fragment specifiers, we can still change the wording or make it dependent on fragment specifiers being present.

**Pull request merged**